go/matrix/matrix.go

```diff
@@ -0,0 +1,100 @@
+// This package implements matrix parsing utilities, to generate arrays
+// from string representations of matrices.
+package matrix
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+)
+
+const ROW_SEPARATOR string = "\n"
+const COLUMN_SEPARATOR string = " "
+
+//  struct representation of a matrix
+type matrix struct {
+	values [][]int
+}
+
+//  New creates a new matrix object based on a string representation of
+//  said matrix. This representation contains one or more non empty
+//  lines, each containing the same number of integer values, separated
+//  by spaces.
+func New(matrixString string) (*matrix, error) {
+	var numCols int = 0
+	var err error = nil
+
+	rows := strings.Split(matrixString, ROW_SEPARATOR)
+	var numRows int = len(rows)
+	var values = make([][]int, numRows)
+
+	for rowInd, row := range rows {
+		row = strings.Trim(row, " ")
+		rowCells := strings.Split(row, COLUMN_SEPARATOR)
+		if rowInd == 0 {
```

Comment:
this logic could be extracted before the for loop (given that you have your helper method) and numCols is initialized there. Afterwards in the loop you iterate over `rows[1:]` and don't need this if-else statement (just the one for consistency).